atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/sweep/asts/bucketingthings/ConsistentOrderingObjectMapperTest.java

```diff
@@ -0,0 +1,148 @@
+            EMPTY_OPTIONAL, BaseEncoding.base64().decode("OikKBfqDYTEyMyQDtoBiQ3Rlc3T7="),
+            NULL_VALUE, BaseEncoding.base64().decode("OikKBfqAYSODYTEyMyQDtvs="),
+            ZERO_INT_VALUE, BaseEncoding.base64().decode("OikKBfqAYSODYTEyM8CAYkN0ZXN0+w=="));
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource("exampleObjectAndSerialisedJsonForm")
+    public void jsonMapperEmitsSyntacticallyMatchingJson(TestObject exampleObject, String expectedJson)
+            throws JsonProcessingException {
+        assertThat(JSON_OBJECT_MAPPER.writeValueAsString(exampleObject)).isEqualTo(expectedJson);
+    }
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource("exampleObjectAndSerialisedSmileForm")
+    public void smileMapperEmitsConsistentSmile(TestObject exampleObject, byte[] expectedSmile)
+            throws JsonProcessingException {
+        assertThat(ConsistentOrderingObjectMapper.OBJECT_MAPPER.writeValueAsBytes(exampleObject))
+                .isEqualTo(expectedSmile);
+    }
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource("exampleObjectAndSerialisedSmileForm")
+    public void smileMapperCanDeserialiseOldValues(TestObject exampleObject, byte[] expectedSmile) throws IOException {
+        assertThat(ConsistentOrderingObjectMapper.OBJECT_MAPPER.readValue(expectedSmile, TestObject.class))
+                .isEqualTo(exampleObject);
+    }
+
+    @Test
+    public void smileMapperCanIgnoreUnknownValues() throws IOException {
+        SuperSetTestObject superSetTestObject = ImmutableSuperSetTestObject.builder()
+                .a(true)
+                .a123(123)
+                .b("test")
+                .newValueToIgnore(456)
+                .build();
+
+        TestObject expectedTestObject =
+                ImmutableTestObject.builder().a(true).a123(123).b("test").build();
+
+        byte[] smile = ConsistentOrderingObjectMapper.OBJECT_MAPPER.writeValueAsBytes(superSetTestObject);
+        TestObject deserialised = ConsistentOrderingObjectMapper.OBJECT_MAPPER.readValue(smile, TestObject.class);
+        assertThat(deserialised).isEqualTo(expectedTestObject);
+    }
+
+    private static Stream<Arguments> exampleObjectAndSerialisedJsonForm() {
+        return GOLDEN_JSON.entrySet().stream().map(entry -> Arguments.of(entry.getKey(), entry.getValue()));
+    }
+
+    private static Stream<Arguments> exampleObjectAndSerialisedSmileForm() {
+        return GOLDEN_SMILE.entrySet().stream().map(entry -> Arguments.of(entry.getKey(), entry.getValue()));
+    }
+
+    @JsonSerialize(as = ImmutableTestObject.class)
+    @JsonDeserialize(as = ImmutableTestObject.class)
+    @Value.Immutable
+    interface TestObject {
+        // Ordered intentionally non-alphabetically to avoid a case where it happens to be ordered by declaration order
+        // for a given run.
+        @Nullable
+        String b();
+
+        Optional<Boolean> a();
+
+        int a123();
```

Comment:
The test object needs to have a Map I think, to exercise `SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS`.